File: android/apollo/src/test/java/io/muun/apollo/domain/utils/FeeWindowTest.kt
package io.muun.apollo.domain.utils

import io.muun.apollo.BaseTest
import io.muun.apollo.domain.model.FeeWindow
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.within
import org.junit.Test
import org.threeten.bp.ZonedDateTime

class FeeWindowTest: BaseTest() {

    val singleFeeWindow = createFeeWindow(1 to 5.6)
    val someFeeWindow = createFeeWindow(2 to 2.3, 5 to 7.2, 20 to 18.7)

    private val margin = 0.000001

    @Test
    fun `returns the fastest fee rate`() {
        assertThat(someFeeWindow.fastestFeeInSatoshisPerByte).isEqualTo(2.3)
    }

    @Test
    fun `returns the only fee rate as fastest`() {
        assertThat(singleFeeWindow.fastestFeeInSatoshisPerByte).isEqualTo(5.6)
    }

    @Test
    fun `returns the exact target as closest, if present`() {
        val targetedFees = someFeeWindow.targetedFees

        for ((target, expectedRate) in targetedFees.entries) {
            val rate = someFeeWindow.getMinimumFeeInSatoshisPerByte(target)

            assertThat(rate).isCloseTo(expectedRate, within(margin))
        }
    }

    @Test
    fun `returns the closest lower target`() {
        assertThat(someFeeWindow.getMinimumFeeInSatoshisPerByte(4))
            .isCloseTo(2.3, within(margin))

        assertThat(someFeeWindow.getMinimumFeeInSatoshisPerByte(15))
            .isCloseTo(7.2, within(margin))

        assertThat(someFeeWindow.getMinimumFeeInSatoshisPerByte(22))
            .isCloseTo(18.7, within(margin))
    }

    @Test
    fun `returns the lowest target by default`() {
        assertThat(someFeeWindow.getMinimumFeeInSatoshisPerByte(1))
            .isCloseTo(2.3, within(margin))
    }

    @Test
    fun `returns the only fee rate as closest`() {
        for (i in listOf(1, 6, 18, 24)) {
            assertThat(singleFeeWindow.getMinimumFeeInSatoshisPerByte(i))
                .isCloseTo(5.6, within(margin))
        }
    }

    @Test(expected = IllegalArgumentException::class)
    fun `fails check when confirmation target is 0`() {
        someFeeWindow.getMinimumFeeInSatoshisPerByte(0)
    }

    @Test(expected = IllegalArgumentException::class)
    fun `fails check when confirmation target is negative`() {
        someFeeWindow.getMinimumFeeInSatoshisPerByte(-1)
    }

    private fun createFeeWindow(vararg targetToRate: Pair<Int, Double>) =
        FeeWindow(1, ZonedDateTime.now(), mapOf(*targetToRate), 1, 1, 1)
}

File: libwallet/operation/fee_window_test.go
package operation

import "testing"

var singleFeeWindow = &FeeWindow{
	TargetedFees: func() map[uint]float64 {
		fees := make(map[uint]float64)
		fees[1] = 5.6
		return fees
	}(),
}

var someFeeWindow = &FeeWindow{
	TargetedFees: func() map[uint]float64 {
		fees := make(map[uint]float64)
		fees[2] = 2.3
		fees[5] = 7.2
		fees[20] = 18.7
		return fees
	}(),
}

func TestFastestFeeRate(t *testing.T) {

	testCases := []struct {
		desc            string
		feewindow       *FeeWindow
		expectedFeeRate float64
	}{
		{
			desc:            "returns the fastest fee rate",
			feewindow:       someFeeWindow,
			expectedFeeRate: 2.3,
		},
		{
			desc:            "returns the only fee rate as fastest",
			feewindow:       singleFeeWindow,
			expectedFeeRate: 5.6,
		},
	}

	for _, tC := range testCases {
		t.Run(tC.desc, func(t *testing.T) {

			feeRate := tC.feewindow.FastestFeeRate()

			if feeRate != tC.expectedFeeRate {
				t.Fatalf("expected feeRate = %v, got %v", tC.expectedFeeRate, feeRate)
			}
		})
	}
}

func TestMinimumFeeRate(t *testing.T) {
	testCases := []struct {
		desc            string
		feewindow       *FeeWindow
		confTarget      uint
		expectedFeeRate float64
	}{
		{
			desc:            "returns the exact target as closest, if present (1)",
			feewindow:       someFeeWindow,
			confTarget:      2,
			expectedFeeRate: 2.3,
		},
		{
			desc:            "returns the exact target as closest, if present (2)",
			feewindow:       someFeeWindow,
			confTarget:      5,
			expectedFeeRate: 7.2,
		},
		{
			desc:            "returns the exact target as closest, if present (3)",
			feewindow:       someFeeWindow,
			confTarget:      20,
			expectedFeeRate: 18.7,
		},
		{
			desc:            "returns the closest lower target (1)",
			feewindow:       someFeeWindow,
			confTarget:      4,
			expectedFeeRate: 2.3,
		},
		{
			desc:            "returns the closest lower target (2)",
			feewindow:       someFeeWindow,
			confTarget:      15,
			expectedFeeRate: 7.2,
		},
		{
			desc:            "returns the closest lower target (3)",
			feewindow:       someFeeWindow,
			confTarget:      22,
			expectedFeeRate: 18.7,
		},
		{
			desc:            "returns the lowest target by default",
			feewindow:       someFeeWindow,
			confTarget:      1,
			expectedFeeRate: 2.3,
		},
		{
			desc:            "returns the only fee rate as closest (1)",
			feewindow:       singleFeeWindow,
			confTarget:      1,
			expectedFeeRate: 5.6,
		},
		{
			desc:            "returns the only fee rate as closest (2)",
			feewindow:       singleFeeWindow,
			confTarget:      6,
			expectedFeeRate: 5.6,
		}, {
			desc:            "returns the only fee rate as closest (3)",
			feewindow:       singleFeeWindow,
			confTarget:      18,
			expectedFeeRate: 5.6,
		}, {
			desc:            "returns the only fee rate as closest (4)",
			feewindow:       singleFeeWindow,
			confTarget:      24,
			expectedFeeRate: 5.6,
		},
	}

	for _, tC := range testCases {
		t.Run(tC.desc, func(t *testing.T) {

			feeRate, err := tC.feewindow.MinimumFeeRate(tC.confTarget)

			if err != nil {
				t.Fatal(err)
			}
			if feeRate != tC.expectedFeeRate {
				t.Fatalf("expected feeRate = %v, got %v", tC.expectedFeeRate, feeRate)
			}
		})
	}
}

func TestInvalidConfirmationTargets(t *testing.T) {
	testCases := []struct {
		desc       string
		feewindow  *FeeWindow
		confTarget uint
	}{
		{
			desc:       "fails check when confirmation target is 0",
			feewindow:  someFeeWindow,
			confTarget: 0,
		},
	}

	for _, tC := range testCases {
		t.Run(tC.desc, func(t *testing.T) {

			_, err := tC.feewindow.MinimumFeeRate(tC.confTarget)

			if err == nil {
				t.Fatalf("expected test to error")
			}
		})
	}
}

func TestNextHighestBlock(t *testing.T) {
	block := someFeeWindow.NextHighestBlock(10.0)
	if block != 2 {
		t.Fatalf("expected block to be 2, got %v", block)
	}
}
